Added Feature Support  for External URL prefix

varshith257 opened the pull request:
## Which problem is this PR solving?
- <!-- Example: Resolves #123 -->
Resolves #5157 
## Description of the changes
 The changes include:

- Added a new field` UIBasePath` to the `StaticAssetsHandlerOptions` struct.
- I created a new `assetsFS` instance using the `UIBasePath` field if it is non-empty.
- Replaced the `BasePath ` with the `UIBasePath` field.

## How was this change tested?
-  Done

## Checklist
- [ ] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [ ] I have signed all commits
- [ ] I have added unit tests for the new functionality
- [ ] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

yurishkuro commented:
what testing did you do?

varshith257 commented:
Thank you for your question. To verify the fix , I followed these steps:

To verify the functionality, I set up a test environment with a similar traffic flow as described in the issue. I then used the updated NewStaticAssetsHandler function in the Jaeger query service with the new` --query.ui-base-path ` option to allow a different external prefix. I accessed the Jaeger UI through the reverse proxy and checked that the index.html file was loaded correctly. I also inspected the network tab in the browser's developer tools to verify that the static files were loaded correctly with the correct prefix based on the` query_base_path` and `--query.base-path` options. Finally, I tested the Jaeger UI functionality to ensure it was working as expected.

Based on these tests, I can confirm that the functionality has been updated to support different internal and external prefixes, as described in the issue. The updated code should address the issue and allow for more flexibility in setting up the Jaeger UI with a reverse proxy.

Used the --query.ui-base-path option to specify the external prefix for the Jaeger UI. For example, if you want the external prefix to be /viewer, you can run the Jaeger query service with the following command:

`jaeger-query --query.base-path=/jaeger/query --query.ui-base-path=/viewer`

varshith257 commented:
@yurishkuro  Tests have been updated to static_handler_test.

varshith257 commented:
Need your review and guidance to make any changes to it if it's required

yurishkuro commented:
you should make changes to examples/reverse-proxy/docker-compose.yml to illustrate usage of different prefixes

yurishkuro commented on go.mod:
```diff
@@ -98,7 +98,7 @@ require (
 	github.com/eapache/queue v1.1.0 // indirect
 	github.com/elastic/elastic-transport-go/v8 v8.4.0 // indirect
-	github.com/fatih/color v1.14.1 // indirect
+	github.com/fatih/color v1.15.0 // indirect
```
this is not required for this PR

varshith257 replied:
resolved

varshith257 commented:
Sure I will make a commit incorporating the suggestions made

varshith257 commented:
@yurishkuro Done changes as you suggested.
Need your review and guidance to make any changes if it's required

yurishkuro commented:
please list what kind of testing steps you did

varshith257 commented:
@yurishkuro

1.  I added a new test function, `TestRegisterStaticHandlerWithUIBasePath`, to test the case where `UIBasePath` is set to a non-empty value. This test function checks that assets are served with the correct base path and that the UI base path is set correctly in the` index.html` file.
2. I added another new test function, `TestRegisterStaticHandlerWithoutUIBasePath`, to test the case where `UIBasePath` is not set.

yurishkuro commented:
You misunderstand - I am not asking about test functions, I can see those in the PR. I am asking about which manual tests did you perform to verify that your changes are achieving the goal. For example, did you run & test the modified docker-compose file?

varshith257 commented:
I will update on it asap

Sidhartha-Biswal commented:
Hi @varshith257,

I tried with your changes and query.ui-base-path cli flag implementation in flags.go but it did not work. 
Please find the below configuration and error screenshot.
QUERY_BASE_PATH: /jaeger
QUERY_UI_BASE_PATH: /hub/jaeger
![jaegeruierror](https://github.com/jaegertracing/jaeger/assets/119291884/b90182b0-945f-4f8a-93d0-360f8410e9bb)

The static path is still the same as shown below.
![staticpath](https://github.com/jaegertracing/jaeger/assets/119291884/2e9e39cc-6186-4dcc-ae4b-76b7b45710bc)
The QUERY_UI_BASE_PATH is not getting reflected anywhere in the UI(index.html)
What is the relation between QUERY_BASE_PATH and QUERY_UI_BASE_PATH?

You have appended both the basepath and uibasepath in the below line. Could you please tell the reason?
https://github.com/jaegertracing/jaeger/blob/19f43c4b0ac5100eaf041308f1544018d6fc8c28/cmd/query/app/static_handler_test.go#L201

varshith257 commented:
@Sidhartha-Biswal I have also observed this behavior and working on it to fix this

Sidhartha-Biswal commented:
Hi @varshith257,

Could you please tell the status and timeline to implement the fix for the issue?

varshith257 commented:
> Hi @varshith257,
> 
> Could you please tell the status and timeline to implement the fix for the issue?
@yurishkuro, 
I wanted to inform you that due to the sudden start of my exams, I have been inactive for the past few days. I apologize for any inconvenience this may have caused.

manisha5464 commented:
Hi @varshith257,

I was checking through the changes u made in static_Handler.go file. I have few questions.
The default value of query.static-files is an empty string. So the staticAssetsRoot is also an "". In that case what is the use of join function below ?
What is the idea behind joining the staticAssetsRoot and uiBasePath ??

![IMG-20240306-WA0002](https://github.com/jaegertracing/jaeger/assets/71115468/daa2fc56-b4bb-487f-9b68-0eedecb28e94)


@yurishkuro can you please review the code changes and suggest some way forward.

yurishkuro commented on go.mod:
```diff
@@ -141,13 +141,13 @@ require (
 	github.com/mattn/go-isatty v0.0.17 // indirect
 	github.com/mitchellh/copystructure v1.2.0 // indirect
-	github.com/mitchellh/go-testing-interface v1.0.0 // indirect
+	github.com/mitchellh/go-testing-interface v0.0.0-20171004221916-a61a99592b77 // indirect
```
dependency changes are not relevant to this PR

yurishkuro commented on examples/reverse-proxy/docker-compose.yml:
```diff
@@ -10,6 +10,7 @@ services:
     environment:
       # This tells Jaeger that it should run the UI with a given prefix.
       QUERY_BASE_PATH: /jaeger/prefix
+      QUERY_UI_BASE_PATH: /hub/jaeger-query/v1/jaeger/viewer
```
how was this tested?

varshith257 replied:
@yurishkuro Testing the changes. Till now I had an issue with docker and I resolved it and testing it. I will update the  progress of this testing of my code-changes

varshith257 replied:
@yurishkuro When I tried to access this with the new updated UIBASEPATH : 

![image](https://github.com/jaegertracing/jaeger/assets/117595548/70c74a5c-69a6-4383-927c-e752ed85cd01)


IMO I have made related changes to access it through uibasepath and load files as intended. Am I missing anything related?

Can please review my code changes?

manisha5464 commented:
Hi @yurishkuro ,
Can you please describe what is the relation between QUERY_BASE_PATH and QUERY_UI_BASE_PATH?
Can we set both the parameters at the same time?
Like for example QUERY_BASE_PATH: /jaeger and QUERY_UI_BASE_PATH: /hub/jaeger, then which path should take the priority?
Which path should reflect in index.html?

I tested with the changes in this PR  but still the issue exists and not able to load the static assets in Jaeger UI

Can you please provide some more inputs on this requirement that would help to proceed further.
Do you think the changes made below are correct ?

I was checking through the changes made by @varshith257 in static_Handler.go file. I have few questions.
The default value of query.static-files is an empty string. So the staticAssetsRoot is also an "". In that case what is the use of join function below ?
What is the idea behind joining the staticAssetsRoot and uiBasePath ??




![image](https://github.com/jaegertracing/jaeger/assets/71115468/c8c135ef-f622-46e9-b7e0-594ff18039c9)


@yurishkuro , please share some inputs on how the QUERY_UI_BASE_PATH needs to be implemented exactly ?

yurishkuro commented:
> Can you please describe what is the relation between QUERY_BASE_PATH and QUERY_UI_BASE_PATH?

it is explained in the ticket

varshith257 commented:
@manisha5464 The join function in static_Handler.go used to construct the complete URL path for serving static assets. Even though the default value of query.static-files is an empty string, the join function ensures that, when uiBasePath is specified, it is correctly appended to the staticAssetsRoot. This helps in forming the correct URL for serving static assets based on the provided configuration. 

The idea behind this is to ensure that, regardless of the initial value of staticAssetsRoot, the uiBasePath is properly incorporated into the final URL, facilitating the correct serving of static assets for the Jaeger UI.

yurishkuro commented on cmd/query/app/flags.go:
```diff
@@ -70,7 +70,8 @@ type QueryOptionsStaticAssets struct {
 // QueryOptionsBase holds configuration for query service shared with jaeger(v2)
 type QueryOptionsBase struct {
 	// BasePath is the base path for all HTTP routes
-	BasePath string
+	BasePath   string
+	UIBasePath string
```
given the amount of confusion this is already causing, it's a good idea to provide clear documentation for these fields

yurishkuro commented on cmd/query/app/static_handler.go:
```diff
@@ -49,6 +49,7 @@ var (
 func RegisterStaticHandler(r *mux.Router, logger *zap.Logger, qOpts *QueryOptions, qCapabilities querysvc.StorageCapabilities) {
 	staticHandler, err := NewStaticAssetsHandler(qOpts.StaticAssets.Path, StaticAssetsHandlerOptions{
 		BasePath:            qOpts.BasePath,
+		UIBasePath:          qOpts.UIBasePath, // Added UIBasePath here
```
you resolved the commend - how was it addressed?

yurishkuro commented on cmd/query/app/static_handler.go:
```diff
@@ -87,9 +89,12 @@ type loadedConfig struct {
 	assetsFS := ui.StaticFiles
 	if staticAssetsRoot != "" {
-		assetsFS = http.Dir(staticAssetsRoot)
+		uiBasePath := options.UIBasePath
```
ok, it's pointless for me to review beyond this point until you write clear definition what each of the two params means and and how it should be used. The code below is wrong.

yurishkuro commented on cmd/query/app/flags.go:
```diff
@@ -70,7 +70,9 @@ type QueryOptionsStaticAssets struct {
 	// BasePath is the base path for all HTTP routes
-	BasePath string
+	BasePath   string
+	// UIBasePath is the base path for Jaeger UI
```
This comment does not clarify anything, I can infer from the variable name that is a "path for UI". I recommend adding a readme with a diagram (using mermaid markup, or perhaps just a table) that explains the difference. Participants in the diagram are browser, a reverse proxy, and query-service, and at each node show what the URL looks like and which config option is supposed to match that.
